### DISCUSSION TITLE
Dependant selects: Select showing all data instead of specific related data

### DISCUSSION THREAD

**Discussion by fabianwurk** [filamentphp/filament#6424] [Help] [answered]
I have 3 Resources:

- Customers: Add 'Customer' name.
- Domains: Select Customer from Select list. And add 'Domain' name.
- Hostings: Select Customer from Select list. Select 'Domain' name from Select list. 

In Hostings, **I am looking to show only the Domains belonging to a specific selected Customer**, but currently its showing all domains on the Select list:

<img width="1201" alt="Screenshot 2023-05-05 at 13 13 12" src="https://user-images.githubusercontent.com/18347097/236454783-9060a59b-a12f-4911-8cb2-7cef3d152411.png">

In my HostingResource I have the following 'Dependant Select':

```
                Select::make('customer_id')
                    ->label('Customer')
                    ->options(Customer::all()->pluck('name', 'id')->toArray())
                    ->required()
                    ->reactive()
                    ->afterStateUpdated(fn (callable $set) => $set('domain_id', null)),

                Select::make('domain_id')
                    ->label('Domain')
                    ->required()
                    ->options(function (callable $get) {
                        $customer = Customer::find($get('country_id'));
                        if (!$customer) {
                            return Domain::all()->pluck('name', 'id');
                        }
                        return $customer->domains->pluck('name', 'id');
                    })
                    ->reactive(),
```
**The Hostings DB table:**

<img width="531" alt="Screenshot 2023-05-05 at 13 17 08" src="https://user-images.githubusercontent.com/18347097/236455884-4b6ec05c-dd11-41b6-bc0e-5dbcdb9707ec.png">

**The Customers DB table:**

<img width="434" alt="Screenshot 2023-05-05 at 13 19 29" src="https://user-images.githubusercontent.com/18347097/236456078-d1c8daba-29db-4dc3-898f-f9bf1bbe3cec.png">

**The Domains DB table:**

<img width="576" alt="Screenshot 2023-05-05 at 13 20 15" src="https://user-images.githubusercontent.com/18347097/236456200-0c2eeab7-7505-4194-84fe-a8ce22a38c81.png">

**In Hosting Model:**

```
class Hosting extends Model
{
    use HasFactory;

    protected $fillable = [
        'customer_id',
        'domain_id',
        'is_active'
    ];
 
    public function customer()
    {
        return $this->belongsTo(Customer::class);
    }

    public function domains()
    {
        return $this->belongsTo(Domain::class);
    }
}
```

**In Customer Model:**

```
class Customer extends Model
{
    use HasFactory;

    protected $fillable = ['name'];

    public function domains(): HasMany
    {
        return $this->hasMany(Domain::class, 'customer');
    }

    public function hostings()
    {
        return $this->hasMany(Hosting::class);
    }
}
```

**In Domain Model:**

```
class Domain extends Model
{
    use HasFactory;

    protected $fillable = [
        'customer_id',
        'name'
    ];

    public function customer()
    {
        return $this->belongsTo(Customer::class);
    }
}
```

**And I have created RelationManagers for Customer and Domain:**

<img width="263" alt="Screenshot 2023-05-05 at 13 24 18" src="https://user-images.githubusercontent.com/18347097/236457021-9d5e916a-c9b1-469f-8e51-272dea4783c9.png">

Just wondering if I am missing something?

Thanks in advance for any help or guidance as my knowledge of Laravel and Filament is basic.

**Comment by danharrin** [comment]
`Customer::find($get('country_id'));` should be `Customer::find($get('customer_id'));`

**Reply by danharrin** [reply]
Also you can save on memory by not hydrating models:

```
if (!$customer) {
    return Domain::query()->pluck('name', 'id');
}
return $customer->domains()->pluck('name', 'id');
```

**Comment by fabianwurk** [comment]
Thanks Dan. I did notice they 'country_id' just before your reply 🙈. Unfortunately, throwing an error now:

<img width="1135" alt="Screenshot 2023-05-05 at 14 19 03" src="https://user-images.githubusercontent.com/18347097/236468795-53c2aa52-3cb1-4c23-95be-a007522c5026.png">

My knowledge of Laravel is pretty limited so maybe I'm going wrong somewhere else on Domains setup?

<img width="1410" alt="Screenshot 2023-05-05 at 14 18 35" src="https://user-images.githubusercontent.com/18347097/236469189-1ee365f6-739c-4c07-99f0-b16932b55a5c.png">

**Reply by danharrin** [reply]
Update the domains relationship, you dont need to define a `customer` column that doesnt exist:

```
public function domains(): HasMany
    {
        return $this->hasMany(Domain::class);
    }
```

**Reply by fabianwurk** [reply]
Thanks so much for replies Dan. 

I have added that hasMany function to Domain Model (I assume thats where you meant?). 

But could you also please clarify this point:

"you dont need to define a customer `column` that doesnt exist:"

I can't see where its defining that customer column. (And apologies for the newbie questions - I am kinda learning Laravel alongside Filament at the mo).

**Reply by danharrin** [reply, marked as answer]
That HasMany goes on the customer model, it replaces your old one

Previously, you had passed `customer` as the second parameter of the HasMany. That parameter is used to tell Laravel which column should hold the customer ID. Since you have a `customer_id`, Laravel can guess that and you dont need to use a second parameter at all.

**Reply by fabianwurk** [reply]
Dan thanks so much for the solution, and your clarification. That all worked perfectly.

Many thanks
